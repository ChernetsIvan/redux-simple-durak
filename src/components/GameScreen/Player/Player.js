import React from "react";
import PropTypes from "prop-types";

import InvisibleCard from "./../Card/InvisibleCard";
import Card from "./../Card/Card";

//Классы-контейнеры кнопок
import PlayerTakeButtonContainer from "./../../../containers/GameScreen/Player/Buttons/PlayerTakeButtonContainer";
import AiTakeButtonContainer from "./../../../containers/GameScreen/Player/Buttons/AiTakeButtonContainer";
import RemoveCardsButtonContainer from "./../../../containers/GameScreen/Player/Buttons/RemoveCardsButtonContainer";

class Player extends React.Component {
  constructor(props) {
    super(props);

    this.onPrevClick = this.onPrevClick.bind(this);
    this.onNextClick = this.onNextClick.bind(this);

    this.state = {
      startInd: 0,
      endInd: 9
    };
  }

  onPrevClick() {
    this.setState((prevState)=>({
      startInd: prevState.startInd - 1,
      endInd: prevState.endInd - 1
    }));
  }

  onNextClick() {
    this.setState((prevState)=>({
      startInd: prevState.startInd + 1,
      endInd: prevState.endInd + 1
    }));
  }

  render() {
    let cards = this.props.cards.map(el => {
      return (
        <Card
          id={el.id}
          key={el.id}
          rank={el.rank}
          suit={el.suit}
          onCardClick={this.props.onCardClick}
          bootStrapColClass="col-1"
          hidden={false}
        />
      );
    });

    if (cards.length < 10) {
      this.setState({
        startInd: 0,
        endInd: 9
      });
    }

    let output = null;
    if (this.props.cards.length > 0) {
      //Показ кнопок Prev и Next
      let out_PrevButton = null;
      if (this.state.startInd > 0 && this.props.cards.length > 10) {
        out_PrevButton = (
          <button onClick={this.onPrevClick} className="btn btn-info">
            <b>&larr;{this.state.startInd}</b>
          </button>
        );
      }

      let out_NextButton = null;
      if (
        this.state.endInd < this.props.cards.length - 1 &&
        this.props.cards.length > 10
      ) {
        out_NextButton = (
          <button onClick={this.onNextClick} className="btn btn-info">
            <b>{this.props.cards.length - 1 - this.state.endInd}&rarr;</b>
          </button>
        );
      }

      //Выводим максимум 10 карт: от startInd до endInd
      let outCards = [];
      if (this.props.cards.length > 10) {
        outCards = cards.slice(
          this.state.startInd,
          this.state.endInd + 1
        );
      } else {
        outCards = cards;
      }

      output = (
        <div className="container">
          <div className="row justify-content-center">
            <div className="col-1 align-self-center">{out_PrevButton}</div>
            {outCards}
            <div className="col-1 align-self-center">{out_NextButton}</div>
          </div>
        </div>
      );
    } else {
      output = <InvisibleCard />;
    }
    return (
      <div className="container">
        <div className="row">
          <div className="col-1" />
          <div className="col-10">{output}</div>
          <div className="col-1" />
        </div>
        <div className="row justify-content-center">
          <div className="col-auto">
            <RemoveCardsButtonContainer />
            <AiTakeButtonContainer />
            <PlayerTakeButtonContainer />
          </div>
        </div>
      </div>
    );
  }
}

export default Player;

Player.propTypes = {
  cards: PropTypes.array.isRequired,
  onCardClick: PropTypes.func.isRequired
};
