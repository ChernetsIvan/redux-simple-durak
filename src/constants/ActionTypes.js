export const START_GAME_CLICKED = 'START_GAME_CLICKED';
export const BEGIN_GAME_CLICKED = 'BEGIN_GAME_CLICKED';
