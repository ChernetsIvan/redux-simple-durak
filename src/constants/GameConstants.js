export const cardRanks = [
  { text: "6", cardValue: 1 },
  { text: "7", cardValue: 2 },
  { text: "8", cardValue: 3 },
  { text: "9", cardValue: 4 },
  { text: "10", cardValue: 5 },
  { text: "В", cardValue: 6 },
  { text: "Д", cardValue: 7 },
  { text: "К", cardValue: 8 },
  { text: "Т", cardValue: 9 }
];

export const cardSuits = [
  { suit: "Ч" },
  { suit: "К" },
  { suit: "П" },
  { suit: "Б" }
];

export const statusMessages = {
  PlayerAttack: "Ваш ход!",
  PlayerDefence: "Защищайтесь!",
  PlayerDiscard: "Беру! Подбрасывать будете?",
  PlayerLose: "Вы проиграли! Сыграем ещё раз ? :)",
  PlayerWin: "Вы выиграли! Поздравляю :)",
  GameResultDraw: "Ничья!"
};
