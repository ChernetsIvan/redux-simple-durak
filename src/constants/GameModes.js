export const PlayerAttack = "PlayerAttack";
export const PlayerDefence = "PlayerDefence";
export const PlayerDiscard = "PlayerDiscard"; //AI забирает: игрок подбрасывает ему карты
export const AiAttack = "AiAttack";
export const AiDefence = "AiDefence";
export const AiDiscard = "AiDiscard"; //Игрок забирает: AI подбрасывает ему карты

export const PlayerWin = "PlayerWin";
export const AIWin = "AIWin";
export const TheDraw = "TheDraw"; //Ничья