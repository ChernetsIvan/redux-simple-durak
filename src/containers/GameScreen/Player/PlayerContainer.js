import { connect } from 'react-redux';
import Player from './../../../components/GameScreen/Player/Player';
import {clickOnPrev, clickOnNext, clickOnCard} from './../../../actions';

const mapStateToProps = store => ({
    cards: store.game.playerCards,
    playerStartInd: store.game.playerStartInd,
    playerEndInd: store.game.playerEndInd    
});

const mapDispatchToProps =  dispatch => ({   
    onPrevClick: () => {
        dispatch(clickOnPrev());
    },
    onNextClick: () => {
        dispatch(clickOnNext());
    },
    onCardClick: (cardId) => {
        dispatch(clickOnCard(cardId));
    }
});

export default connect(mapStateToProps, mapDispatchToProps)(Player);