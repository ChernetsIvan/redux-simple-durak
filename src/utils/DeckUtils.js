import { pull, find } from 'lodash-es';

import {cardRanks, cardSuits} from './../constants/GameConstants';

import CardModel from './../models/CardModel';

class DeckUtils{
    //отсортировать входной массив карт по возрастанию isAsc=true
    //(убыванию-> isAsc=false)  по параметру 'power'
    static sortInputDeckByPower(inputDeck, isAsc){
        inputDeck.sort(function(a,b){
            if(isAsc){
                return a.power - b.power;
            }else{
                return b.power - a.power;
            }        
        });
    }

    //Возвращает наполненную всеми 36-ю картами колоду
    static fillCards(trumpSuit){
        let fullDeck = [];
        //Устанавливает power для каждой карты по принципу:
        //козырь->10*rank.cardValue, НЕкозырь->1*rank.cardValue
        let z = 0;
        let power, coef;               
        cardSuits.forEach(function(element,index, array){
            if(element.suit === trumpSuit.suit){
                coef = 10;
            }else{
                coef = 1;
            }
            cardRanks.forEach(function(element2, index2, array2){
                power = coef * element2.cardValue;
                fullDeck.push(
                    new CardModel(z, element2, element, power)
                );
                z++;
            });
        });
        return fullDeck;
    }

    //Возвращает перемешанную колоду
    static shuffle(inputDeck){
        var newFullDeck = [];
        while(inputDeck.length > 0){
            //Случайный индекс в inputDeck
            let randIndex = Math.random() * (inputDeck.length-1);
            randIndex = Math.round(randIndex);

            let card = inputDeck[randIndex];
            newFullDeck.push(card);

            inputDeck.splice(randIndex, 1);
        }
        return newFullDeck;
    }

    //Перемещает первую попавшуюся карту с козырной мастью
    //в конец колоды 
    static moveAnyCardWithTrumpSuitToTailOfFullDeck(trumpSuit, fullDeck){
        let card = find(fullDeck, (card) => {
            return card.suit.suit === trumpSuit.suit;
        });
        pull(fullDeck, card);
        fullDeck.push(card);
    }

    //взять из НАЧАЛА (!) колоды fullDeck и положить в inputDeck столько карт, 
    //чтобы в inputDeck в итоге было 6 карт
    static giveUpToSixCards(fullDeck, inputDeck){
        //если хотя бы одно из условий не выполняется, в цикл не заходим:
        while(inputDeck.length < 6 && fullDeck.length > 0){
            let card = fullDeck.shift();
            inputDeck.push(card);
        }
    }

    //Возвращает козырную масть (не карту!) для будущей игры
    static chooseSuitForTrumpCardInFuture(){
        let randIndex = Math.random() * (cardSuits.length-1);
        randIndex = Math.round(randIndex);
        return cardSuits[randIndex];
    }
}

export default DeckUtils;