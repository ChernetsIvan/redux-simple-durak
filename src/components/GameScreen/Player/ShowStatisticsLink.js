import React from "react";
import PropTypes from "prop-types";
import { Link } from "react-router";

import * as gameModes from "./../../../constants/GameModes";

const ShowStatisticsLink = ({ gameMode }) => {
  let dash = <span>&mdash;</span>;
  return gameMode === gameModes.AIWin ||
  gameMode === gameModes.PlayerWin ||
  gameMode === gameModes.TheDraw ? (
    <Link
      to="/statistics"
      className="btn btn-secondary ml-2"
    >
      {dash} Игра закончена, смотреть статистику!
    </Link>
  ) : null;
};

export default ShowStatisticsLink;

ShowStatisticsLink.propTypes = {
  gameMode: PropTypes.string.isRequired
};
