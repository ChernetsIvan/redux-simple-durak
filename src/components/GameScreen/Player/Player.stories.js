import React from "react";

import { storiesOf } from "@storybook/react";
import { action } from "@storybook/addon-actions";

import { Provider } from "react-redux";
import { createStore } from "redux";

import Player from "./Player";
import DeckUtils from "./../../../utils/DeckUtils";
import * as gameModes from "./../../../constants/GameModes";

import reducer from "./../../../reducers";

const store = createStore(reducer);

const cards = DeckUtils.fillCards({ suit: "Ðš" });

const clickOnCard = action("Card was clicked!");

storiesOf("Player", module)
  .addDecorator(story => <Provider store={store}>{story()}</Provider>)
  .add("with 4 cards", () => (
    <Player cards={cards.slice(10, 14)} onCardClick={clickOnCard} />
  ))
  .add("with 7 cards", () => (
    <Player cards={cards.slice(28, 35)} onCardClick={clickOnCard} />
  ))
  .add("with 12 cards", () => (
    <Player cards={cards.slice(0, 12)} onCardClick={clickOnCard} />
  ))
  .add("with 36 cards", () => (
    <Player cards={cards.slice(0, 36)} onCardClick={clickOnCard} />
  ));
