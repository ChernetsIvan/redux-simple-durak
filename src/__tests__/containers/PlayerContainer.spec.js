import React from "react";
import renderer from "react-test-renderer";

import { Provider } from "react-redux";
import { createStore } from "redux";

import PlayerContainer from "./../../containers/GameScreen/Player/PlayerContainer";
import reducer from "./../../reducers";
import * as actionTypes from "./../../constants/ActionTypes";

describe("PlayerContainer snapshot testing", () => {
  let store;

  beforeEach(() => {
    store = createStore(reducer);
  });

  test("Store in initial-state mode", () => {
    const component = renderer.create(
      <Provider store={store}>
        <PlayerContainer />
      </Provider>
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  test.skip("Dispatch with START_GAME_CLICKED-action", () => {
    store.dispatch({
      type: actionTypes.START_GAME_CLICKED
    });
    const component = renderer.create(
      <Provider store={store}>
        <PlayerContainer />
      </Provider>
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
});



/*
import React from "react";
import renderer from "react-test-renderer";

import DeckUtils from "./../../utils/DeckUtils";
import Player from "./../../components/GameScreen/Player/Player";

describe("Player snapshot testing", () => {  
  const emptyFunc = () => {};
  const cards = DeckUtils.fillCards({ suit: "Ðš" });

  test("with count of cards > 10", () => {
    const component = renderer.create(
      <Player cards={cards} onCardClick={emptyFunc} />
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  test("with count of cards < 10", () => {
    const component = renderer.create(
      <Player cards={cards.slice(0,5)} onCardClick={emptyFunc} />
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
});
*/