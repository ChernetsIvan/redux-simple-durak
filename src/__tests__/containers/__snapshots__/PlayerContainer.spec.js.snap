// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlayerContainer snapshot testing Dispatch with START_GAME_CLICKED-action 1`] = `
<div
  className="container"
>
  <div
    className="row"
  >
    <div
      className="col-1"
    />
    <div
      className="col-10"
    >
      <div
        className="container"
      >
        <div
          className="row justify-content-center"
        >
          <div
            className="col-1 align-self-center"
          />
          <div
            className="col-1 card mb-1"
            onClick={[Function]}
          >
            <div
              className="row"
            >
              <b
                className="ml-1"
              >
                <b
                  className="redText"
                >
                  6
                </b>
              </b>
            </div>
            <div
              className="row"
            >
              <img
                alt=""
                className="littleSuitImg ml-1"
                src="heart.png"
              />
            </div>
            <div
              className="row justify-content-center"
            >
              <img
                alt=""
                className="bigSuitImg my-2"
                src="heart.png"
              />
            </div>
            <div
              className="row justify-content-end"
            >
              <img
                alt=""
                className="littleSuitImg rotated mr-1"
                src="heart.png"
              />
            </div>
            <div
              className="row justify-content-end"
            >
              <b
                className="rotated mr-1"
              >
                <b
                  className="redText"
                >
                  6
                </b>
              </b>
            </div>
          </div>
          <div
            className="col-1 card mb-1"
            onClick={[Function]}
          >
            <div
              className="row"
            >
              <b
                className="ml-1"
              >
                <b
                  className="blackText"
                >
                  10
                </b>
              </b>
            </div>
            <div
              className="row"
            >
              <img
                alt=""
                className="littleSuitImg ml-1"
                src="spade.png"
              />
            </div>
            <div
              className="row justify-content-center"
            >
              <img
                alt=""
                className="bigSuitImg my-2"
                src="spade.png"
              />
            </div>
            <div
              className="row justify-content-end"
            >
              <img
                alt=""
                className="littleSuitImg rotated mr-1"
                src="spade.png"
              />
            </div>
            <div
              className="row justify-content-end"
            >
              <b
                className="rotated mr-1"
              >
                <b
                  className="blackText"
                >
                  10
                </b>
              </b>
            </div>
          </div>
          <div
            className="col-1 card mb-1"
            onClick={[Function]}
          >
            <div
              className="row"
            >
              <b
                className="ml-1"
              >
                <b
                  className="blackText"
                >
                  Д
                </b>
              </b>
            </div>
            <div
              className="row"
            >
              <img
                alt=""
                className="littleSuitImg ml-1"
                src="spade.png"
              />
            </div>
            <div
              className="row justify-content-center"
            >
              <img
                alt=""
                className="bigSuitImg my-2"
                src="spade.png"
              />
            </div>
            <div
              className="row justify-content-end"
            >
              <img
                alt=""
                className="littleSuitImg rotated mr-1"
                src="spade.png"
              />
            </div>
            <div
              className="row justify-content-end"
            >
              <b
                className="rotated mr-1"
              >
                <b
                  className="blackText"
                >
                  Д
                </b>
              </b>
            </div>
          </div>
          <div
            className="col-1 card mb-1"
            onClick={[Function]}
          >
            <div
              className="row"
            >
              <b
                className="ml-1"
              >
                <b
                  className="redText"
                >
                  К
                </b>
              </b>
            </div>
            <div
              className="row"
            >
              <img
                alt=""
                className="littleSuitImg ml-1"
                src="heart.png"
              />
            </div>
            <div
              className="row justify-content-center"
            >
              <img
                alt=""
                className="bigSuitImg my-2"
                src="heart.png"
              />
            </div>
            <div
              className="row justify-content-end"
            >
              <img
                alt=""
                className="littleSuitImg rotated mr-1"
                src="heart.png"
              />
            </div>
            <div
              className="row justify-content-end"
            >
              <b
                className="rotated mr-1"
              >
                <b
                  className="redText"
                >
                  К
                </b>
              </b>
            </div>
          </div>
          <div
            className="col-1 card mb-1"
            onClick={[Function]}
          >
            <div
              className="row"
            >
              <b
                className="ml-1"
              >
                <b
                  className="blackText"
                >
                  Т
                </b>
              </b>
            </div>
            <div
              className="row"
            >
              <img
                alt=""
                className="littleSuitImg ml-1"
                src="spade.png"
              />
            </div>
            <div
              className="row justify-content-center"
            >
              <img
                alt=""
                className="bigSuitImg my-2"
                src="spade.png"
              />
            </div>
            <div
              className="row justify-content-end"
            >
              <img
                alt=""
                className="littleSuitImg rotated mr-1"
                src="spade.png"
              />
            </div>
            <div
              className="row justify-content-end"
            >
              <b
                className="rotated mr-1"
              >
                <b
                  className="blackText"
                >
                  Т
                </b>
              </b>
            </div>
          </div>
          <div
            className="col-1 card mb-1"
            onClick={[Function]}
          >
            <div
              className="row"
            >
              <b
                className="ml-1"
              >
                <b
                  className="blackText"
                >
                  10
                </b>
              </b>
            </div>
            <div
              className="row"
            >
              <img
                alt=""
                className="littleSuitImg ml-1"
                src="club.png"
              />
            </div>
            <div
              className="row justify-content-center"
            >
              <img
                alt=""
                className="bigSuitImg my-2"
                src="club.png"
              />
            </div>
            <div
              className="row justify-content-end"
            >
              <img
                alt=""
                className="littleSuitImg rotated mr-1"
                src="club.png"
              />
            </div>
            <div
              className="row justify-content-end"
            >
              <b
                className="rotated mr-1"
              >
                <b
                  className="blackText"
                >
                  10
                </b>
              </b>
            </div>
          </div>
          <div
            className="col-1 align-self-center"
          />
        </div>
      </div>
    </div>
    <div
      className="col-1"
    />
  </div>
  <div
    className="row justify-content-center"
  >
    <div
      className="col-auto"
    />
  </div>
</div>
`;

exports[`PlayerContainer snapshot testing Store in initial-state mode 1`] = `
<div
  className="container"
>
  <div
    className="row"
  >
    <div
      className="col-1"
    />
    <div
      className="col-10"
    >
      <div
        className="container invisible"
      >
        <div
          className="row"
        >
          <div
            className="col-1 card mb-1 shirt"
          />
        </div>
      </div>
    </div>
    <div
      className="col-1"
    />
  </div>
  <div
    className="row justify-content-center"
  >
    <div
      className="col-auto"
    />
  </div>
</div>
`;
